src/vm/scanning.rs

```diff
@@ -31,6 +32,41 @@ impl<F: FnMut(ObjectReference) -> ObjectReference> ObjectTracer for F {
     }
 }
 
+/// A `QueuingTracerFactory` can create a "tracer" object local to a GC Worker to trace and
+/// update edges, and create `ScanObjects` work packets to expand the transitive closure starting
+/// from newly traced objects.  It is used during root scanning and binding-side weak reference
+/// processing.
+pub trait QueuingTracerFactory<VM: VMBinding>: Clone + Send + 'static {
```

Comment:
A "tracer" (i.e. `ObjectTracer`) provides the `trace_object` method.  "Queuing" means adding the object into a buffer if it is first visited during a transitive closure computing.  While most use cases of `trace_object` have this queuing behaviour, `LXRConcurrentTraceObjects` behaves differently because it enqueues the *children* of the visited object (See https://github.com/wenyuzhao/mmtk-core/blob/lxr/src/plan/lxr/cm.rs#L148).  But it is still a different style of enqueuing.

I think it should be safe to assume that the `trace_object` always has some kind of queuing behaviour, therefore we can omit the word "queuing" from its name.

The word "Factory" here looks odd, too, for the reason you described, although it actually creates `ObjectTracer`.  I think the name `ObjectTracerContext` is better.  `ScopedTracer` may be another option, although a `ScopedTracer` itself is not a tracer.  p.s. I originally designed it like a factory that returns `ObjectTracer` instances, but that'll need the GAT feature of newer Rust and it'll need manual flushing.

Reply:
> I am not quite comfortable about adding a new type for each API method.

I introduced `ProcessWeakRefsContext` for a purpose.  My concern is that if we extend the `process_weak_refs` method and provide more information (in addition to `forwarding` and `is_nursery`) through its parameters, existing VM bindings will need to update their source code for the added parameters even if they don't use the new information.  With a `ProcessWeakRefsContext` struct, we can safely add new fields just in case some VMs need more information.

For example, if we add a parameter:
```diff
-fn process_weak_refs(forwarding: bool) -> bool;
+fn process_weak_refs(forwarding: bool, nursery: bool) -> bool;
```

Existing VM bindings will need to add the parameter even if it is not using that information.

```diff
 impl Collection for OpenJDK {
-    fn process_weak_refs(forwarding: bool) -> bool {
+    fn process_weak_refs(forwarding: bool, _nursery: bool) -> bool {
 // ... more code here
```

But if we provide a struct in the first place, we just add fields to that struct.

```diff
 struct ProcessWeakRefsContext
     forwarding: bool,
+    nursery: bool,
 }
 trait Collection {
     fn process_weak_refs(context: ProcessWeakRefsContext) -> bool; // This doesn't change.
```

> Do you have an idea what other API methods may use this type? Can this type be used in anyway when a binding wants to process node/edges immediately in stack root scanning?

Yes.  It will be used for root scanning as well.  I'll make another PR to extend the root scanning API (the `RootsWorkFactory` trait) so that it can provide the functionality of `ObjectTracerContext`.  I'll define it like `trait RootsWorkFactory : ObjectTracerContext` so that we can reuse this `ObjectTracerContext` trait.

> My general preference is to add methods to existing types, and make them more general. I am exactly sure if that is applicable here.

There is a risk.  This is exactly why we had the `TransitiveClosure` trait 15 years ago, because Daniel wanted to introduce concurrent mark-sweep, and wanted to make a general interface (See https://wks.github.io/blog/2022/05/16/fifteen-years-transitiveclosure.html#but-why-did-we-migrate-away-from-it).

Reply:
If it will be reused, then I am okay with it. 

About `ProcessWeakRefsContext`, yeah, that is the same idea as I tidied up the arguments for creating plans/spaces. I am wondering what other arguments we may pass to a binding in the context struct. One concern of mine is abstraction leakage. I don't feel that we want to the binding to know too much about intimate knowledge on the plan/GC algorithm. I know we have `plan.is_current_gc_nursery()` which is not good either.

Reply:
I renamed this trait to `ObjectTracerContext`, and its method to just `with_tracer`.

I also managed to reuse the code in `ScanObjects` around the call to `Scanning::scan_object_and_trace_edges`.  They happen to have the same logic, and they both use the `ObjectTracer` trait.